packages/@sanity/import/src/import.js

```diff
@@ -0,0 +1,65 @@
+const debug = require('debug')('sanity:import')
+const flatten = require('lodash/flatten')
+const validateOptions = require('./validateOptions')
+const streamToArray = require('./streamToArray')
+const {getAssetRefs, unsetAssetRefs} = require('./assetRefs')
+const assignArrayKeys = require('./assignArrayKeys')
+const uploadAssets = require('./uploadAssets')
+const batchDocuments = require('./batchDocuments')
+const importBatches = require('./importBatches')
+const {
+  getStrongRefs,
+  weakenStrongRefs,
+  setTypeOnReferences,
+  strengthenReferences
+} = require('./references')
+
+async function importFromStream(stream, opts) {
+  const options = validateOptions(stream, opts)
+
+  // Get raw documents from the stream
+  debug('Streaming input source to array of documents')
+  options.onProgress({step: 'Reading/validating data file'})
+  const raw = await streamToArray(stream)
+
+  // User might not have applied `_key` on array elements which are objects;
+  // if this is the case, generate random keys to help realtime engine
+  const keyed = raw.map(doc => assignArrayKeys(doc))
+
+  // Sanity prefers to have a `_type` on every object. Make sure references
+  // has `_type` set to `reference`.
+  const docs = keyed.map(doc => setTypeOnReferences(doc))
+
+  // Find references that will need strengthening when import is done
+  const strongRefs = docs.map(getStrongRefs).filter(Boolean)
+
+  // Extract asset references from the documents
+  const assetRefs = flatten(docs.map(getAssetRefs).filter(ref => ref.length))
+
+  // Remove asset references from the documents
+  const assetless = docs.map(unsetAssetRefs)
+
+  // Make strong references weak so they can be imported in any order
+  const weakened = assetless.map(weakenStrongRefs)
+
+  // Create batches of documents to import. Try to keep batches below a certain
+  // byte-size (since document may vary greatly in size depending on type etc)
+  const batches = batchDocuments(weakened)
+
```

Comment:
Its not clear to me, so just checking:  `importBatches`, `uploadAssets` and `strengthenReferences` must be done in sequence, right?

Reply:
Sort of. `importBatches` must be done first, but `uploadAssets` and `strengthenReferences` could be done in parallel. The reason why we're not doing this at the moment is that if one of these operations fail, there is no way to stop the other. I've made a note of it in the readme as a definite improvement for the future, but it would require using something like a queue instead of simply mapping over items with a concurrency.

Reply:
I see, thanks for the clarification! And +1 for improved error handling in the future. If `strengthenReferences` fail right now, it will result in a lot of dangling assets too, or are the asset uploads idempotent?

Reply:
Asset uploads are done first and connected to their corresponding documents. If this operation fails, it will have dangling assets, however. When re-running the import, it should find the same assets already uploaded and reuse those, however.

Failures on strengthened references will also leave some references as weak, and this is a situation which might easily occur if you are refering to documents which do not exist (incorrect IDs or similar). I've also made a note of this in the readme, that we should attempt to check if referenced documents actually exist before starting upload.

All in all, "rolling back" a failed upload is pretty hard.

Reply:
> When re-running the import, it should find the same assets already uploaded and reuse those, however.

This is great. I have no further comments 👩‍⚖️

Reply:
> This is great. I have no further comments :woman_judge:

Achievement unlocked.